features/graph/service/src/main/java/org/opennms/netmgt/graph/service/DefaultGraphContainerCache.java

```diff
@@ -0,0 +1,131 @@
+/*******************************************************************************
+ * This file is part of OpenNMS(R).
+ *
+ * Copyright (C) 2019-2019 The OpenNMS Group, Inc.
+ * OpenNMS(R) is Copyright (C) 1999-2019 The OpenNMS Group, Inc.
+ *
+ * OpenNMS(R) is a registered trademark of The OpenNMS Group, Inc.
+ *
+ * OpenNMS(R) is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published
+ * by the Free Software Foundation, either version 3 of the License,
+ * or (at your option) any later version.
+ *
+ * OpenNMS(R) is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with OpenNMS(R).  If not, see:
+ *      http://www.gnu.org/licenses/
+ *
+ * For more information contact:
+ *     OpenNMS(R) Licensing <license@opennms.org>
+ *     http://www.opennms.org/
+ *     http://www.opennms.com/
+ *******************************************************************************/
+
+package org.opennms.netmgt.graph.service;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import org.opennms.netmgt.graph.api.ImmutableGraphContainer;
+import org.opennms.netmgt.graph.api.service.GraphContainerCache;
+
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+
+public class DefaultGraphContainerCache implements GraphContainerCache {
```

Comment:
I wonder why you implemented the locking and caching "by hand" instead of using something like a ConcurrentHashMap? Maybe you want to add a comment about the reasons behind?

An alternative would maybe be to use Guava Caching? https://github.com/google/guava/wiki/CachesExplained
That could be an easy and elegant solution which offers lots of possibilities out of the box?

Reply:
Ideally we would use the `CacheBuilder` from guava. However periodically invalidating only SOME of the cache values is not supported by guava. That is why I implemented it myself.

Reply:
Take a look at the Caffeine cache - based on Guava with more features. We use it for caching DNS responses with varying TTLs: https://github.com/OpenNMS/opennms/blob/develop/features/dnsresolver/netty/src/main/java/org/opennms/netmgt/dnsresolver/netty/CaffeineCache.java#L61

Reply:
Cool, I can take a look :)

Reply:
Caffeine cache is providing the features we need to make this work. Nice 👍